lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -138,59 +96,105 @@ def common_checksum_type(agent_checksum_type)
+            basedir_meta = list_of_data.find {|meta| meta.relative_path == '.'}
+            devfail "FileServing::Metadata search should always return the root search path" if basedir_meta.nil?
+
+            if ! basedir_meta.full_path.start_with? environment_path.to_s
+              # If any source is not in the environment path, skip inlining this resource.
+              sources_in_environment = false
+              break
+            end
 
-          child_resource = Puppet::Resource.new(:file, File.join(file[:path], meta.relative_path))
-          child_resource[:source] = meta.source
-          replace_metadata(child_resource, meta)
+            base_content_uri = get_content_uri(basedir_meta, source, environment_path)
+            list_of_data.each do |metadata|
+              if metadata.relative_path == '.'
+                metadata.content_uri = base_content_uri
+              else
+                metadata.content_uri = "#{base_content_uri}/#{metadata.relative_path}"
+              end
+            end
 
-          # Copy parameters from original parent directory
-          file.original_parameters.each do |param, value|
-            # These should never be passed to our children
-            unless [:parent, :ensure, :recurse, :recurselimit, :target, :alias, :source].include? param
-              child_resource[param] = value.to_s
+            source_to_metadatas[source] = list_of_data
+            # Optimize for returning less data if sourceselect is first
+            if resource[:sourceselect] == 'first' || resource[:sourceselect].nil?
+              break
             end
           end
-          child_resource
-        end.compact
-        child_resources = Puppet::Type::File.remove_less_specific_files(child_resources, resource[:path] || resource.title, list_of_resources) do |file_resource|
-          file_resource[:path] || file_resource.title
         end
 
-        #TODO: Add edge between `child_resource` and its nearest ancestor and
-        # add an edge between `child_resource` and all other resources that depend
-        # on `file`
-        child_resources.each do |child|
-          catalog.add_resource(child)
-          catalog.add_edge(file, child)
+        if sources_in_environment && !source_to_metadatas.empty?
+          catalog.recursive_metadata[resource.title] = source_to_metadatas
         end
       else
-        metadata = file.parameter(:source).metadata
+        options = {
+          :environment        => catalog.environment_instance,
+          :links              => resource[:links] ? resource[:links].to_sym : :manage,
+          :checksum_type      => resource[:checksum] ? resource[:checksum].to_sym : checksum_type.to_sym,
+          :source_permissions => resource[:source_permissions] ? resource[:source_permissions].to_sym : :ignore
+        }
+
+        metadata = nil
+        sources.each do |source|
+          if data = Puppet::FileServing::Metadata.indirection.find(source, options)
```

Comment:
Also, we've added a `checksum_value` parameter to the file type, so you can specify the desired checksum type and value, and puppet will only attempt to retrieve the file if the checksum doesn't match, eliminating the need to retrieve http metadata:

```puppet
file { '/tmp/foo':
  ensure   => file,
  checksum => md5,
  checksum_value => '4e9531911f7a8caef3f20198457ba89b',
  source   => 'https://raw.githubusercontent.com/puppetlabs/puppet/master/Gemfile',
}
```